addons/account/tests/test_account_invoice_rounding.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+
+from odoo.addons.account.tests.account_test_users import AccountTestUsers
+from odoo.exceptions import ValidationError
+
+import time
+
+
+class TestAccountInvoiceRounding(AccountTestUsers):
+
+    def setUp(self):
+        super(AccountTestUsers, self).setUp()
+        self.account_receivable = self.env['account.account'].search(
+            [('user_type_id', '=', self.env.ref('account.data_account_type_receivable').id)], limit=1)
+        self.account_revenue = self.env['account.account'].search(
+            [('user_type_id', '=', self.env.ref('account.data_account_type_revenue').id)], limit=1)
+
+    def create_cash_rounding(self, rounding, method, strategy):
+        return self.env['account.cash.rounding'].create({
+            'name': 'rounding ' + method,
+            'rounding': rounding,
+            'account_id': self.account_receivable.id,
+            'strategy': strategy,
+            'rounding_method': method,
+        })
+
+    def create_invoice(self, amount, cash_rounding_id, tax_amount=None):
+        """ Returns an open invoice """
+        invoice_id = self.env['account.invoice'].create({
+            'partner_id': self.env.ref("base.res_partner_2").id,
+            'reference_type': 'none',
+            'currency_id': self.env.ref('base.USD').id,
+            'name': 'invoice test rounding',
+            'account_id': self.account_receivable.id,
+            'type': 'out_invoice',
+            'date_invoice': time.strftime('%Y') + '-06-26',
+        })
+        tax_id = None
+        if tax_amount:
+            tax_id = self.env['account.tax'].search([('name', '=', 'Test Tax')])
+            if tax_id:
+                tax_id.amount = tax_amount
+            else:
+                tax_id = self.env['account.tax'].create({
+                    'name': 'Test Tax',
+                    'amount': tax_amount,
+                    'amount_type': 'fixed',
+                })
+        self.env['account.invoice.line'].create({
+            'product_id': self.env.ref("product.product_product_4").id,
+            'quantity': 1,
+            'price_unit': amount,
+            'invoice_id': invoice_id.id,
+            'name': 'something',
+            'account_id': self.account_revenue.id,
+            'invoice_line_tax_ids': [(6, 0, [tax_id.id])] if tax_id else None
+        })
+        # Create the tax_line_ids
+        invoice_id._onchange_invoice_line_ids()
+
+        # We need to set the cash_rounding_id after the _onchange_invoice_line_ids
+        # to avoid a ValidationError from _check_cash_rounding because the onchange
+        # are not well triggered in the tests.
+        try:
+            invoice_id.cash_rounding_id = cash_rounding_id
+        except ValidationError:
+            pass
```

Comment:
wtf?

Reply:
In a perfect world, the normal behavior will be:
- assign cash_rounding_id
- trigger onchange (_onchange_cash_rounding)
- trigger constrains (_check_cash_rounding)

... but during tests, the ORM bypass the onchange step and then, the ValidationError in _check_cash_rounding is triggered.